cmd/publicize/server.go

```diff
@@ -0,0 +1,154 @@
+package main
+
+import (
+	"fmt"
+	"regexp"
+	"strings"
+
+	"github.com/sirupsen/logrus"
+
+	prowconfig "k8s.io/test-infra/prow/config"
+	"k8s.io/test-infra/prow/git/v2"
+	"k8s.io/test-infra/prow/github"
+	"k8s.io/test-infra/prow/pluginhelp"
+	"k8s.io/test-infra/prow/plugins"
+)
+
+var publicizeRe = regexp.MustCompile(`(?mi)^/publicize\s*$`)
+
+func helpProvider(_ []prowconfig.OrgRepo) (*pluginhelp.PluginHelp, error) {
+	pluginHelp := &pluginhelp.PluginHelp{
+		Description: `The publicize plugin is used for merging a pull request into the bonded repository.`,
+	}
+	pluginHelp.AddCommand(pluginhelp.Command{
+		Usage:       "/publicize",
+		Description: "Merge the specific pull request into the bonded repository",
+		WhoCanUse:   "Members of the trusted organization for the repo.",
+		Examples:    []string{"/publicize"},
+	})
+	return pluginHelp, nil
+}
+
+type server struct {
+	githubTokenGenerator func() []byte
+
+	gitName  string
+	gitEmail string
+
+	bonds map[string]string
+
+	ghc github.Client
+	gc  git.ClientFactory
+}
+
+func (s *server) handleIssueComment(l *logrus.Entry, ic github.IssueCommentEvent) {
+	if !publicizeRe.MatchString(ic.Comment.Body) {
+		return
+	}
+
+	org := ic.Repo.Owner.Login
+	repo := ic.Repo.Name
+	num := ic.Issue.Number
+	commentAuthor := ic.Comment.User.Login
+
+	logger := logrus.WithFields(logrus.Fields{
+		github.OrgLogField:  org,
+		github.RepoLogField: repo,
+		github.PrLogField:   num,
+	})
+
+	if !ic.Issue.IsPullRequest() {
+		s.createComment(ic, fmt.Sprintf("Publicize plugin can only be used in pull requests"), logger)
+		return
+	}
+
+	// Only org members should be able to publicize a pull request.
+	ok, err := s.ghc.IsMember(org, commentAuthor)
+	if err != nil {
+		return
+	}
+	if !ok {
+		s.createComment(ic, fmt.Sprintf("only [%s](https://github.com/orgs/%s/people) org members may request a publicize", org, org), logger)
+		return
+	}
+	logger.Info("Publicize of PR has been requested.")
+
+	pr, err := s.ghc.GetPullRequest(org, repo, num)
+	if err != nil {
+		return
+	}
+	baseBranch := pr.Base.Ref
+
+	if !pr.Merged {
+		s.createComment(ic, "cannot publicize an unmerged PR", logger)
+		return
+	}
+
+	logger.Info("Searching bond...")
+	bondedOrgRepo, ok := s.bonds[fmt.Sprintf("%s/%s", org, repo)]
+	if !ok {
+		logger.Warn("There is no public repository configured for the current repository.")
+		s.createComment(ic, "cannot publicize because there is no bond for the current pull request", logger)
+		return
+	}
+
+	bondedOrg := strings.Split(bondedOrgRepo, "/")[0]
+	bondedRepo := strings.Split(bondedOrgRepo, "/")[1]
+
+	logger.Infof("Trying to merge the PR to destination: %s/%s@%s", bondedOrg, bondedRepo, baseBranch)
+	if err := s.mergeAndPushToRemote(bondedOrg, bondedRepo, org, repo, baseBranch); err != nil {
+		logger.WithError(err).Warnf("couldn't merge the pull request and push to the destination: %v", err)
+		s.createComment(ic, fmt.Sprintf("Publicize failed with error: %v", err), logger)
```

Comment:
Are we sure that the `err` cannot leak a token possibly embedded in the URL? Are we sure that the error is censored?

Reply:
I was having nightmares with this. I tried also to simulate some errors but I didn't see any leaked token. I will dig this further. Also it would be easier to debug this once I finish with the tests as well.

Reply:
Just run the secret agent's censor on the comment body before posting

Reply:
I didn't know that this is possible. I'll check that. Thanks.